### PR TITLE
Issue #224 Introducing a MINISHIFT_HOME env variable

### PR DESCRIPTION
**Pull request by hferentschik**

### CODE REVIEW

**Comment by hferentschik** [comment]
WDYT? So far I have not added any docs yet, since I really just need this for now for where I am heading with (integration) testing, but maybe we could add this somewhere as tip?

**Comment by LalatenduMohanty** [comment]
@hferentschik With the current code a different home for .minishift can be set usng MINISHIFT_HOME environment variable. But are you also planning to add a command line flag for this?

**Comment by LalatenduMohanty** [comment]
May be we can add documentation for this PR In https://github.com/minishift/minishift/blob/master/docs/using.md

File: pkg/minikube/constants/constants.go [hunk 1]
```diff
@@ -68,11 +54,29 @@ const (
 	RemoteOpenShiftCAPath  = "/var/lib/minishift/openshift.local.config/master/ca.crt"
 )
 
-var ConfigFilePath = MakeMiniPath("config")
```
Comment (praveenkumar) [inline pkg/minikube/constants/constants.go]: @LalatenduMohanty yes this we can use for integration purpose I think. Not used currently.

Comment (hferentschik) [inline pkg/minikube/constants/constants.go]: Right

**Comment by hferentschik** [comment]
> With the current code a different home for .minishift can be set usng MINISHIFT_HOME environment variable.

Yes. Driving force are (integration) testing. Can of course be used by users as well, provided we want to make it "public"

> But are you also planning to add a command line flag for this?

I was not planning to for now. I don't think this will be so easy and needs potentially further refactoring.

> May be we can add documentation for this

I can find a spot somewhere. It was more a question whether we want to do this at this stage or not.

**Comment by praveenkumar** [comment]
> Yes. Driving force are (integration) testing. Can of course be used by users as well, provided we want to make it "public"

I think to make this change public it might required some more refactoring and testing so I will say lets use it for integration purpose right now and the we plan it for public.

**Comment by LalatenduMohanty** [comment]
@hferentschik @praveenkumar  Here is my 2 cents. Looks like a proper implementation need more time and resource. Also any code that gets committed in in the project is anyway public. That being said I am fine with documenting that this feature is only used for integration testing and users are not supposed to use it.

**Comment by praveenkumar** [comment]
I tried to test this PR but looks like due to network (might be it's slow at my home) it failed. I will try to do test it again on Monday.

```
$ make integration
CGO_ENABLED=0 GOARCH=amd64 GOOS=darwin go build --installsuffix cgo -ldflags="-X github.com/minishift/minishift/pkg/version.version=1.0.0-beta.1 -X github.com/minishift/minishift/pkg/version.openshiftVersion=v1.3.1 -s -w -extldflags '-static'" -o ./out/darwin-amd64/minishift ./cmd/minikube
go test -timeout 3600s github.com/minishift/minishift/test/integration --tags=integration

EXTRA_ARGS='
--label provider=virtualbox
--insecure-registry 172.30.0.0/16

'
CACERT=/var/lib/boot2docker/ca.pem
DOCKER_HOST='-H tcp://0.0.0.0:2376'
DOCKER_STORAGE=aufs
DOCKER_TLS=auto
SERVERKEY=/var/lib/boot2docker/server-key.pem
SERVERCERT=/var/lib/boot2docker/server.pem

export "FOO=BAR"
export "BAZ=BAT"


--- FAIL: TestStartStop (381.78s)
	util.go:44: Error running command: start --vm-driver virtualbox E0107 17:57:07.533767   35420 start.go:131] Error starting machine: Error creating. Error creating machine: Error checking the host: Error checking and/or regenerating the certs: There was an error validating certificates for host "192.168.99.100:2376": dial tcp 192.168.99.100:2376: i/o timeout
		You can attempt to regenerate them using 'docker-machine regenerate-certs [name]'.
		Be advised that this will trigger a Docker daemon restart which might stop running containers.
		. Retrying.
		E0107 18:01:19.071214   35420 start.go:238] Error starting 'cluster up':  exit status 1
		. Output: Starting local OpenShift cluster...
		Downloading ISO
 36.00 MB / 36.00 MB [=================================================================================================================================================] 100.00% 0s
 16.16 MB / 16.16 MB [=================================================================================================================================================] 100.00% 0s
		Provisioning OpenShift via '/var/folders/lq/ywt_b86d7vzdl7qpscqvp51m0000gp/T/minishift-integration-test-957912516/cache/oc/v1.3.1/oc [cluster up]'
		-- Checking OpenShift client ... OK
		-- Checking Docker client ... OK
		-- Checking Docker version ... OK
		-- Checking for existing OpenShift container ... OK
		-- Checking for openshift/origin:v1.3.1 image ... 
		   Pulling image openshift/origin:v1.3.1
		   Pulled 0/7 layers, 0% complete
		   Pulled 1/7 layers, 14% complete
		   Pulled 2/7 layers, -92459880% complete
		   Pulled 3/7 layers, -270109216% complete
		   Pulled 3/7 layers, -586548928% complete
		   Pulled 3/7 layers, -895504384% complete
		   Pulled 3/7 layers, -1181288192% complete
		   Pulled 4/7 layers, -1034547200% complete
		   Pulled 4/7 layers, -1505704192% complete
		   Pulled 5/7 layers, -532934656% complete
		   Pulled 6/7 layers, 86% complete
		   Pulled 5/7 layers, 71% complete
		FAIL
		   Error: error pulling Docker image openshift/origin:v1.3.1
		   Caused By:
		     Error: Error pulling image (v1.3.1) from docker.io/openshift/origin, Server error: Status 0 while fetching image layer (3690474eb5b4b26fdfbd89c6e159e8cc376ca76ef48032a30fa6aafd56337880)
FAIL
FAIL	github.com/minishift/minishift/test/integration	1541.388s
make: *** [integration] Error 1
```

**Comment by praveenkumar** [comment]
@hferentschik Did testing in office network and integration test passed.

```
mkdir -p ./out/darwin-amd64
CGO_ENABLED=0 GOARCH=amd64 GOOS=darwin go build --installsuffix cgo -ldflags="-X github.com/minishift/minishift/pkg/version.version=1.0.0-beta.1 -X github.com/minishift/minishift/pkg/version.openshiftVersion=v1.3.1 -s -w -extldflags '-static'" -o ./out/darwin-amd64/minishift ./cmd/minikube
go test -timeout 3600s github.com/minishift/minishift/test/integration --tags=integration
ok  	github.com/minishift/minishift/test/integration	736.996s
```

File: pkg/minikube/constants/constants_test.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+func TestMinishiftHomeViaEnvironment(t *testing.T) {
+	expectedMiniPath := "/tmp/minishift-test-home"
+	os.Setenv("MINISHIFT_HOME", expectedMiniPath)
+	defer os.Unsetenv("MINISHIFT_HOME")
+
+	actualMiniPath := getMinishiftHomeDir()
+	if actualMiniPath != expectedMiniPath {
+		t.Fatalf("Expected Minishift home directory : '%s' Got: '%s'", expectedMiniPath, actualMiniPath)
+	}
+}
```
Comment (praveenkumar) [inline pkg/minikube/constants/constants_test.go]: Forget to add new line. (run go fmt)

Comment (hferentschik) [inline pkg/minikube/constants/constants_test.go]: ok

**Comment by hferentschik** [comment]
> Here is my 2 cents. Looks like a proper implementation need more time and resource. 

What do you understand under "proper"? What do you think is missing here? 

> Also any code that gets committed in the project is anyway public. 

Sure, it's public, but that does not mean that it is supported. Documenting something means you are making a commitment to the community. Trying to change this later on requires more thought and effort, than changing something which was never supported. The latter you can just change whenever.

> That being said I am fine with documenting (anywhere) that this feature should only used for integration testing and users are not supposed to use it.

In this case we don't need to document anything. The integration tests are in flux for now.

**Comment by hferentschik** [comment]
> I think to make this change public it might required some more refactoring

What do you have in mind? I am trying to understand what you and @LalatenduMohanty think is missing here

There is no flag for this, but there is a reason for that. How would I read this flag. You would have to read this flag after Viper is initialized, right? At this stage Viper would have already read its persistence config, right? And where would it get that from? If I am supposed to get the minishift home dir from a flag, I need to delay the Viper initialization (at least the persistence config one) to after I read the flag in question, since I will need to read the config from that directory. Might be possible, not sure. If not, one would need to fiddle with the plain flags bypassing Viper. Not so appealing. 

 > and testing so I will say lets use it for integration purpose right now and the we plan it for public.

What's about I am adding a "experimental" note?

**Comment by hferentschik** [comment]
Merged, thanks.

**Pull request merged**